lib/src/game/base_game.dart

```diff
@@ -31,8 +33,35 @@ class BaseGame extends Game with FPSCounter {
   /// Components to be removed on the next update
   final Set<Component> _removeLater = {};
 
-  /// Camera position; every non-HUD component is translated so that the camera position is the top-left corner of the screen.
-  Vector2 camera = Vector2.zero();
+  /// The camera translates the coordinate space after the viewport is applied.
+  final Camera camera = Camera();
+
+  /// The viewport transforms the coordinate space depending on your chosen
+  /// implementation.
+  /// The default implementation no-ops, but you can use this to have a fixed
+  /// screen ratio for example.
+  Viewport viewport = DefaultViewport();
```

Comment:
Why not implement this on Game instead of BaseGame? This seems something that even games that don't use components would benefit from this

Reply:
I'm only reluctant because this basically changes the render method, and on Game the render method was supposed to be totally empty for the user to implement. if they want they can add a viewport to they Game and use it on render. it's 3 extra lines. if they don't want they don't get the burden of it

Reply:
I don't see that as a problem, must callSuper is our friend for that

Reply:
that is exactly the problem, I don't want the user of Game to be forced to call super. Game should be almost like an interface (that dart doesn't have) with empty render and update and you can override and do what you want. it only hooks the bare minimum necessary. does not add anything on top. that is BaseGame, preconfigured for your convenience. you can still use viewport with game, it just won't be included automatically

Reply:
I still think that this should be in the Game class, if you don't want to use mustCallSuper, we could have this code hooked on an internal render method, and that internal render method would call the abstract render method that the user needs to implement. We have done something similar like that in the past

Reply:
we do that if we need to, because it's something coming from the engine that we have to expose somehow. I don't think this is the case here, this is just something that you put on top and get some benefits, it's super easy to wire and on Game you get to wire it however you want (for example you might want to render some UI outside the viewport, etc). The BaseGame has our opinionated suggestion of how to use it. I don't see much advantage in adding it since it's so simple to add yourself, and I would rather keep Game as slim as possible. we should only have things that are strictly necessary there

Reply:
Hmm, on second thought here I agree with luan.
A viewport on there is making it opinionated.

Reply:
If we separate the methods and leave the abstract render without anything, Game will still be slim, and since you said yourself that this is only a few lines to wire this into game, this will not add much to the Game class.

Camera is a basic function IMO, it makes a lot of sense on being on our basic Game class

Reply:
I don't see what exactly do you want in these methods, the methods are already pretty separated on the viewport class, you can just hook it up. there is nothing to add on the Game class.

also we are talking about viewport right? camera is tied to base game as it uses position component

Reply:
I am not talking about viewport exactly, I am talking about this feature as a whole. I think this should be on the Game class, not the BaseGame class.

The method I am talking about is not about the Camera or Viewport classes, but about that with a small change on Game methods, this could easily be wired directly on the Game class without making it "Fat"

Reply:
yes, it could. but it's so easy I don't see the point, any user can wire it up if they want, the way they want. I think we should only have on Game what is crucial and would be able to be added otherwise. like if you want to use Game and use Components you can, just add a list of Components, and handle it the way you like.

Reply:
will tackle this on the followup